    Author: Mikael Koli
    email: koli.mikael@gmail.com
    Date: 21.01.2019
    State: Development

# Demonstration of Finnish Business Portal

This file demonstrates the use of Finnish Business Portal Python API. In order to replicate the demo, please make sure you have appropriate version of Python and the libraries Pandas and Requests and their dependencies. After this, download this application from the source.

## Navigation:
* [Setup](#H0)
  * [Import the module](#H0_imports)
* [Parameters](#H1)
* [Example Searches](#H20)
  * [Simple Search](#H21)
  * [Deep Search](#H22)
* [Example Application](#H3)
  * [Getting the Addresses of Searched Companies](#H31)
  * [Getting the Public Notices of Searched Companies](#H32)

## Setup<a class="anchor" id="H0"></a>

In [1]:
import datetime
import platform
import os
import sys

print(f'Python version: {platform.python_version()}')
print(f'''
Time of Run: 
    Year:    {datetime.datetime.now().year}
    Month:   {datetime.datetime.now().month}
    Day:     {datetime.datetime.now().day}
    Time:    {datetime.datetime.now().time()}
''')

Python version: 3.6.6

Time of Run: 
    Year:    2019
    Month:   1
    Day:     21
    Time:    20:28:27.154134



### Import the module<a class="anchor" id="H0_imports"></a>

In [2]:
import finnish_business_portal as bportal

## Parameters<a class="anchor" id="H1"></a>

The parameters with description for the API can be found using show_search_parameters(). These parameters can be directly used in the search functions as keyword arguments. The Open Data API requires the parameters to be in camelCase but this application turn it from snake_case for you.

In [3]:
bportal.show_search_parameters()

Seach Parameters
---------------------------------------------------------------------
max_results:                Maximum results to show
results_from:               Number of result to start with
name:                       name of the company of to search
business_id:                Business ID (Finnish: Y-tunnus)
registered_office:          City where the organization is registered
street_address_post_code:   Visiting address
company_form:               Form of the company
business_line:              Industry name
business_line_code:         Industry code
company_registration_from:  Registeration date
---------------------------------------------------------------------



In addition, there are different registers in the Open Data API. These are different from other parameters as they are defined as arguments in the API's URL thus are not shown in the search parameters. However, this does not affect in the use of this application in any other way. You can get the registers and descriptions using show_registers(). These registers are accessed using keyword argument "register" when calling the search functions.

In [4]:
bportal.show_registers()

Registers
-----------------------------------------------------------------------
tr:   Finnish Trade Register (Finnish: KAUPPAREKISTERIN KUULUTUSTIEDOT)
bis:  Finnish Business Information System (Finnish: YTJ:n tietopankki)
-----------------------------------------------------------------------



## Example Searches<a class="anchor" id="H20"></a>

There are currently two search functions: 
- search_companies
- search_companies_deep


The function search_companies handles all the internet accesses. However, this function returns data on the top level only and does not visit in the detailed URL's even though the Open Data API returns detailed version when queried with only one business ID. The function search_companies_deep handles this issue and loops all of the found business IDs individually resulting in more complete data but longer search.

There is also additional keyword arguments that can be specified in addition to the search parameters and these parameters' names end with under score (ie. wait_time_). Please see the doc strings of the functions for more information.

### Simple Search<a class="anchor" id="H21"></a>

In [5]:
bportal.search_companies(name="Fortum", register="tr", display_="default")

Status: 200
Showing 10 of 15 companies.


bisDetailsUri companyForm  \
businessId                                                                  
2679784-7   http://avoindata.prh.fi/opendata/bis/v1/2679784-7          OY   
2481338-0   http://avoindata.prh.fi/opendata/bis/v1/2481338-0          OY   
2481332-1   http://avoindata.prh.fi/opendata/bis/v1/2481332-1          OY   
2481336-4   http://avoindata.prh.fi/opendata/bis/v1/2481336-4          OY   
2185030-3   http://avoindata.prh.fi/opendata/bis/v1/2185030-3          OY   
2079891-0   http://avoindata.prh.fi/opendata/bis/v1/2079891-0          OY   
1852303-7   http://avoindata.prh.fi/opendata/bis/v1/1852303-7          OY   
1852328-0   http://avoindata.prh.fi/opendata/bis/v1/1852328-0          OY   
1762952-6   http://avoindata.prh.fi/opendata/bis/v1/1762952-6          OY   
1723707-1   http://avoindata.prh.fi/opendata/bis/v1/1723707-1          OY   

                                                  detailsUri  \
businessId                                                     
2679784-7   http://avoindata.prh.fi/opendata/tr/v1/2679784-7   
2481338-0   http://avoindata.prh.fi/opendata/tr/v1/2481338-0   
2481332-1   http://avoindata.prh.fi/opendata/tr/v1/2481332-1   
2481336-4   http://avoindata.prh.fi/opendata/tr/v1/2481336-4   
2185030-3   http://avoindata.prh.fi/opendata/tr/v1/2185030-3   
2079891-0   http://avoindata.prh.fi/opendata/tr/v1/2079891-0   
1852303-7   http://avoindata.prh.fi/opendata/tr/v1/1852303-7   
1852328-0   http://avoindata.prh.fi/opendata/tr/v1/1852328-0   
1762952-6   http://avoindata.prh.fi/opendata/tr/v1/1762952-6   
1723707-1   http://avoindata.prh.fi/opendata/tr/v1/1723707-1   

                                        name registrationDate  
businessId                                                     
2679784-7              Fortum Real Estate Oy       2015-06-30  
2481338-0                      Fortum C&H Oy       2012-06-01  
2481332-1   Fortum Power and Heat Holding Oy       2012-06-01  
2481336-4                     Fortum Norm Oy       2012-06-01  
2185030-3                   Fortum Growth Oy       2008-03-26  
2079891-0           Fortum Asiakaspalvelu Oy       2006-12-14  
1852303-7             Fortum Heat and Gas Oy       2004-05-01  
1852328-0                  Fortum Markets Oy       2004-02-01  
1762952-6                         Fortume Oy       2002-05-08  
1723707-1                        Fortumat Oy       2001-10-04

Function search_companies does not loop the results by default and is showing only the number of companies specified in max_results. To get all of the companies,
<br>a) increase "max_results" (may crash the query if too large) or
<br>b) set "loop_results_" to True

In [6]:
bportal.search_companies(name="Fortum", register="tr", display_="default", loop_results_=True)

Status: 200
Found 15 companies
Looping through...
Progress: 67 %


bisDetailsUri companyForm  \
businessId                                                                  
2679784-7   http://avoindata.prh.fi/opendata/bis/v1/2679784-7          OY   
2481338-0   http://avoindata.prh.fi/opendata/bis/v1/2481338-0          OY   
2481332-1   http://avoindata.prh.fi/opendata/bis/v1/2481332-1          OY   
2481336-4   http://avoindata.prh.fi/opendata/bis/v1/2481336-4          OY   
2185030-3   http://avoindata.prh.fi/opendata/bis/v1/2185030-3          OY   
2079891-0   http://avoindata.prh.fi/opendata/bis/v1/2079891-0          OY   
1852303-7   http://avoindata.prh.fi/opendata/bis/v1/1852303-7          OY   
1852328-0   http://avoindata.prh.fi/opendata/bis/v1/1852328-0          OY   
1762952-6   http://avoindata.prh.fi/opendata/bis/v1/1762952-6          OY   
1723707-1   http://avoindata.prh.fi/opendata/bis/v1/1723707-1          OY   
1526527-1   http://avoindata.prh.fi/opendata/bis/v1/1526527-1          OY   
1463611-4   http://avoindata.prh.fi/opendata/bis/v1/1463611-4         OYJ   
0680422-9   http://avoindata.prh.fi/opendata/bis/v1/0680422-9          OY   
0350017-4   http://avoindata.prh.fi/opendata/bis/v1/0350017-4          OY   
0109160-2   http://avoindata.prh.fi/opendata/bis/v1/0109160-2          OY   

                                                  detailsUri  \
businessId                                                     
2679784-7   http://avoindata.prh.fi/opendata/tr/v1/2679784-7   
2481338-0   http://avoindata.prh.fi/opendata/tr/v1/2481338-0   
2481332-1   http://avoindata.prh.fi/opendata/tr/v1/2481332-1   
2481336-4   http://avoindata.prh.fi/opendata/tr/v1/2481336-4   
2185030-3   http://avoindata.prh.fi/opendata/tr/v1/2185030-3   
2079891-0   http://avoindata.prh.fi/opendata/tr/v1/2079891-0   
1852303-7   http://avoindata.prh.fi/opendata/tr/v1/1852303-7   
1852328-0   http://avoindata.prh.fi/opendata/tr/v1/1852328-0   
1762952-6   http://avoindata.prh.fi/opendata/tr/v1/1762952-6   
1723707-1   http://avoindata.prh.fi/opendata/tr/v1/1723707-1   
1526527-1   http://avoindata.prh.fi/opendata/tr/v1/1526527-1   
1463611-4   http://avoindata.prh.fi/opendata/tr/v1/1463611-4   
0680422-9   http://avoindata.prh.fi/opendata/tr/v1/0680422-9   
0350017-4   http://avoindata.prh.fi/opendata/tr/v1/0350017-4   
0109160-2   http://avoindata.prh.fi/opendata/tr/v1/0109160-2   

                                        name registrationDate  
businessId                                                     
2679784-7              Fortum Real Estate Oy       2015-06-30  
2481338-0                      Fortum C&H Oy       2012-06-01  
2481332-1   Fortum Power and Heat Holding Oy       2012-06-01  
2481336-4                     Fortum Norm Oy       2012-06-01  
2185030-3                   Fortum Growth Oy       2008-03-26  
2079891-0           Fortum Asiakaspalvelu Oy       2006-12-14  
1852303-7             Fortum Heat and Gas Oy       2004-05-01  
1852328-0                  Fortum Markets Oy       2004-02-01  
1762952-6                         Fortume Oy       2002-05-08  
1723707-1                        Fortumat Oy       2001-10-04  
1526527-1                        Fortumex Oy       1999-03-11  
1463611-4                         Fortum Oyj       1998-02-07  
0680422-9                   Fortum Assets Oy       1987-12-09  
0350017-4          Fortum Waste Solutions Oy       1979-12-14  
0109160-2           Fortum Power and Heat Oy       1932-07-05

### Deep Search<a class="anchor" id="H22"></a>

Function search_companies is only useful for acquiring the list of companies but not much else as the data is limited. Use search_companies_deep to gain thorough information from all of the companies. Beware that the search might take time as all of the found business IDs must be queryed individually separately from the registers.

In [7]:
bportal.search_companies_deep(name="Fortum", display_="default")

Time taken at least: 40 seconds
Progress: 10 %
Progress: 20 %
Progress: 30 %
404 Client Error: Not Found for url: http://avoindata.prh.fi/tr/v1?totalResults=true&businessId=2059590-2
404 Client Error: Not Found for url: http://avoindata.prh.fi/tr/v1?totalResults=true&businessId=2059592-9
Progress: 40 %
Progress: 50 %
Progress: 60 %
Progress: 70 %
Progress: 80 %
Progress: 90 %


tr  \
                                                   addresses   
2679784-7  [{'street': 'Keilalahdentie 2-4', 'postCode': ...   
2481332-1  [{'street': 'Keilalahdentie 2-4', 'postCode': ...   
2481336-4  [{'street': 'Keilalahdentie 2-4', 'postCode': ...   
2481338-0  [{'street': 'Keilalahdentie 2-4', 'postCode': ...   
2185030-3  [{'street': 'Keilalahdentie 2-4', 'postCode': ...   
2079891-0  [{'street': 'Keilalahdentie 2-4', 'postCode': ...   
1852328-0  [{'street': 'PL 100', 'postCode': '00048', 'ty...   
1852303-7  [{'street': 'PL 100', 'postCode': '00048', 'ty...   
2059590-2                                                NaN   
2059592-9                                                NaN   

                                                              \
                                              auxiliaryNames   
2679784-7                                                 []   
2481332-1                                                 []   
2481336-4                                                 []   
2481338-0                                                 []   
2185030-3                                                 []   
2079891-0  [{'order': 1, 'name': 'FORTUM ESD', 'registrat...   
1852328-0  [{'order': 1, 'name': 'Fortum E-House', 'regis...   
1852303-7                                                 []   
2059590-2                                                NaN   
2059592-9                                                NaN   

                                                                        \
                                               bisDetailsUri checkDate   
2679784-7  http://avoindata.prh.fi/opendata/bis/v1/2679784-7      None   
2481332-1  http://avoindata.prh.fi/opendata/bis/v1/2481332-1      None   
2481336-4  http://avoindata.prh.fi/opendata/bis/v1/2481336-4      None   
2481338-0  http://avoindata.prh.fi/opendata/bis/v1/2481338-0      None   
2185030-3  http://avoindata.prh.fi/opendata/bis/v1/2185030-3      None   
2079891-0  http://avoindata.prh.fi/opendata/bis/v1/2079891-0      None   
1852328-0  http://avoindata.prh.fi/opendata/bis/v1/1852328-0      None   
1852303-7  http://avoindata.prh.fi/opendata/bis/v1/1852303-7      None   
2059590-2                                                NaN       NaN   
2059592-9                                                NaN       NaN   

                                                                          \
          companyForm                                       companyForms   
2679784-7          OY  [{'type': 'OY', 'registrationDate': '2015-06-3...   
2481332-1          OY  [{'type': 'OY', 'registrationDate': '2012-06-0...   
2481336-4          OY  [{'type': 'OY', 'registrationDate': '2012-06-0...   
2481338-0          OY  [{'type': 'OY', 'registrationDate': '2012-06-0...   
2185030-3          OY  [{'type': 'OY', 'registrationDate': '2008-03-2...   
2079891-0          OY  [{'type': 'OY', 'registrationDate': '2006-12-1...   
1852328-0          OY  [{'type': 'OY', 'registrationDate': '2004-02-0...   
1852303-7          OY  [{'type': 'OY', 'registrationDate': '2004-05-0...   
2059590-2         NaN                                                NaN   
2059592-9         NaN                                                NaN   

                                                                      \
                                                 detailsUri language   
2679784-7  http://avoindata.prh.fi/opendata/tr/v1/2679784-7       FI   
2481332-1  http://avoindata.prh.fi/opendata/tr/v1/2481332-1       FI   
2481336-4  http://avoindata.prh.fi/opendata/tr/v1/2481336-4       FI   
2481338-0  http://avoindata.prh.fi/opendata/tr/v1/2481338-0       FI   
2185030-3  http://avoindata.prh.fi/opendata/tr/v1/2185030-3       FI   
2079891-0  http://avoindata.prh.fi/opendata/tr/v1/2079891-0       FI   
1852328-0  http://avoindata.prh.fi/opendata/tr/v1/1852328-0       FI   
1852303-7  http://avoindata.prh.fi/opendata/tr/v1/1852303-7       FI  

404 Client Error means that a business ID was not found from the register(s). This exception is silenced only in the deep search.

In [8]:
bportal.search_companies_deep(name=["Fortum", "Nokia"], display_="default", max_results=3)

Time taken at least: 24 seconds
Progress: 8 %
Progress: 17 %
Progress: 25 %
Progress: 33 %
Progress: 42 %
Progress: 50 %
Progress: 58 %
Progress: 67 %
Progress: 75 %
Progress: 83 %
Progress: 92 %


tr                 \
                                                    addresses auxiliaryNames   
businessId                                                                     
2679784-7   [{'street': 'Keilalahdentie 2-4', 'postCode': ...             []   
2481332-1   [{'street': 'Keilalahdentie 2-4', 'postCode': ...             []   
2481336-4   [{'street': 'Keilalahdentie 2-4', 'postCode': ...             []   
2947233-5   [{'street': 'Elopolku 3', 'postCode': '37200',...             []   
2892396-6   [{'street': 'Tammikatu 10', 'postCode': '37120...             []   
2889182-9   [{'street': 'Taivalkunnantie 1048', 'postCode'...             []   

                                                                         \
                                                bisDetailsUri checkDate   
businessId                                                                
2679784-7   http://avoindata.prh.fi/opendata/bis/v1/2679784-7      None   
2481332-1   http://avoindata.prh.fi/opendata/bis/v1/2481332-1      None   
2481336-4   http://avoindata.prh.fi/opendata/bis/v1/2481336-4      None   
2947233-5   http://avoindata.prh.fi/opendata/bis/v1/2947233-5      None   
2892396-6   http://avoindata.prh.fi/opendata/bis/v1/2892396-6      None   
2889182-9   http://avoindata.prh.fi/opendata/bis/v1/2889182-9      None   

                                                                           \
           companyForm                                       companyForms   
businessId                                                                  
2679784-7           OY  [{'type': 'OY', 'registrationDate': '2015-06-3...   
2481332-1           OY  [{'type': 'OY', 'registrationDate': '2012-06-0...   
2481336-4           OY  [{'type': 'OY', 'registrationDate': '2012-06-0...   
2947233-5           OY  [{'type': 'OY', 'registrationDate': '2018-11-0...   
2892396-6           OY  [{'type': 'OY', 'registrationDate': '2018-03-1...   
2889182-9           OY  [{'type': 'OY', 'registrationDate': '2018-02-1...   

                                                                       \
                                                  detailsUri language   
businessId                                                              
2679784-7   http://avoindata.prh.fi/opendata/tr/v1/2679784-7       FI   
2481332-1   http://avoindata.prh.fi/opendata/tr/v1/2481332-1       FI   
2481336-4   http://avoindata.prh.fi/opendata/tr/v1/2481336-4       FI   
2947233-5   http://avoindata.prh.fi/opendata/tr/v1/2947233-5       FI   
2892396-6   http://avoindata.prh.fi/opendata/tr/v1/2892396-6       FI   
2889182-9   http://avoindata.prh.fi/opendata/tr/v1/2889182-9       FI   

                                                                      \
           latestRegistrationDate                               name   
businessId                                                             
2679784-7              2018-06-02              Fortum Real Estate Oy   
2481332-1              2018-05-16   Fortum Power and Heat Holding Oy   
2481336-4              2018-06-02                     Fortum Norm Oy   
2947233-5              2018-11-05  Nokian Saneeraus Asiantuntijat Oy   
2892396-6              2018-03-12             Nokian JV-Saneeraus Oy   
2889182-9              2018-02-16              Nokian Timanttityö Oy   

                 ...         \
                 ...          
businessId       ...          
2679784-7        ...          
2481332-1        ...          
2481336-4        ...          
2947233-5        ...          
2892396-6        ...          
2889182-9        ...          

                                                          bis  \
                                                 companyForms   
businessId                                                      
2679784-7   [{'version': 1, 'name': 'Limited company', 'ty...   
2481332-1   [{'version': 1, 'name': 'Limited company', 'ty...   
2481336-4   [{'version': 1, 'name': 'Limited company', '

## Example Application<a class="anchor" id="H3"></a>

In [9]:
import pandas as pd

In [10]:
df = bportal.search_companies_deep(name="Fortum", display_="default", register_="tr")

Time taken at least: 20 seconds
Progress: 10 %
Progress: 20 %
Progress: 30 %
Progress: 40 %
Progress: 50 %
Progress: 60 %
404 Client Error: Not Found for url: http://avoindata.prh.fi/tr/v1?totalResults=true&businessId=2059590-2
Progress: 70 %
404 Client Error: Not Found for url: http://avoindata.prh.fi/tr/v1?totalResults=true&businessId=2059592-9
Progress: 80 %
Progress: 90 %


In [11]:
def flat_column(series):
    "Turn a series containing list of dicts to multi-index dataframe"
    column_name = series.name
    df_partial_flat = series.apply(pd.Series)
    return pd.concat(
        [df_partial_flat[col].apply(pd.Series) for col in df_partial_flat.columns], 
        axis=1, sort=False, 
        keys=[f'{column_name} {n_col}' for n_col in df_partial_flat.columns]
    )


### Getting the Addresses of Searched Companies<a class="anchor" id="H31"></a>

In [13]:
flat_column(df["addresses"])

addresses 0                                        \
                        street postCode type    city country website   
businessId                                                             
2679784-7   Keilalahdentie 2-4    02150    1   Espoo      FI    None   
2481332-1   Keilalahdentie 2-4    02150    1   Espoo      FI    None   
2481336-4   Keilalahdentie 2-4    02150    1   Espoo      FI    None   
2481338-0   Keilalahdentie 2-4    02150    1   Espoo      FI    None   
2185030-3   Keilalahdentie 2-4    02150    1   Espoo      FI    None   
2079891-0   Keilalahdentie 2-4    02150    1   Espoo      FI    None   
1852328-0               PL 100    00048    2  Fortum      FI    None   
1852303-7               PL 100    00048    2  FORTUM      FI    None   

                                                          ... addresses 1  \
               phone         fax registrationDate endDate ...    postCode   
businessId                                                ...               
2679784-7   010 4511        None       2017-12-21    None ...       00048   
2481332-1    0104511        None       2017-12-21    None ...       00048   
2481336-4    0104511        None       2017-12-21    None ...       00048   
2481338-0    0104511        None       2017-12-21    None ...       00048   
2185030-3    0104511        None       2017-12-21    None ...       00048   
2079891-0    0104511  0104532353       2017-12-21    None ...       00048   
1852328-0       None        None       2017-10-13    None ...         NaN   
1852303-7       None        None       2009-02-27    None ...         NaN   

                                                               \
           type    city country website     phone         fax   
businessId                                                      
2679784-7   2.0  Fortum      FI     NaN  010 4511        None   
2481332-1   2.0  Fortum      FI     NaN   0104511        None   
2481336-4   2.0  Fortum      FI     NaN   0104511        None   
2481338-0   2.0  Fortum      FI     NaN   0104511        None   
2185030-3   2.0  Fortum      FI     NaN   0104511        None   
2079891-0   2.0  Fortum      FI     NaN   0104511  0104532353   
1852328-0   NaN     NaN     NaN     NaN       NaN         NaN   
1852303-7   NaN     NaN     NaN     NaN       NaN         NaN   

                                         
           registrationDate endDate   0  
businessId                               
2679784-7        2017-12-21     NaN NaN  
2481332-1        2017-12-21     NaN NaN  
2481336-4        2017-12-21     NaN NaN  
2481338-0        2017-12-21     NaN NaN  
2185030-3        2017-12-21     NaN NaN  
2079891-0        2017-12-21     NaN NaN  
1852328-0               NaN     NaN NaN  
1852303-7               NaN     NaN NaN  

[8 rows x 21 columns]

### Getting the Public Notices of Searched Companies<a class="anchor" id="H32"></a>

In [15]:
flat_column(df["publicNotices"]).unstack().swaplevel(-1, 0).swaplevel(-1, 1).sort_index().dropna()
# Unstacking sets the columns as indexes and with swaplevel method the convenient 
# order of the multi-index is set. Lastly, the index is sorted and non existing data is dropped.

businessId                                     
1852303-7   publicNotices 0  detailsUri            http://avoindata.prh.fi/opendata/tr/v1/publicn...
                             entryCodes                                                       [TASE]
                             recordNumber                                                2018/27466X
                             registrationDate                                             2018-05-16
                             typeOfRegistration                                                   TA
            publicNotices 1  detailsUri            http://avoindata.prh.fi/opendata/tr/v1/publicn...
                             entryCodes                                                       [NIMP]
                             recordNumber                                                2016/733803
                             registrationDate                                             2016-10-04
                             typeOfRegistra

The meaning of "entryCodes" and "typeOfRegistration" can be found from the API's website.
- Entry codes: http://avoindata.prh.fi/tr-codes_v1.fi.txt
- Type of Registration: http://avoindata.prh.fi/tr-type_v1.fi.txt